rfcs/2020-10-15-3480-file-source-rework.md

```diff
@@ -0,0 +1,425 @@
+changes in this RFC that should impact performance positively, and we'll have
+a better view of the potential cost/benefit of these approaches once those have
+landed.
+
+#### Compression-aware fingerprints
+
+As it stands, both `Fingerprinter` and `FileWatcher` instances read data from
+files, one to checksum that data and the other to return lines. The problem is
+that only `FileWatcher` handles compression, so our fingerprints can get
+confused if a file is rotated and compressed.
+
+To address that, we should evolve `FileWatcher` into a more general wrapper
+struct for file handles. This will allow us to encapsulate all direct file
+access to within the struct, where we can more easily ensure that concerns like
+compression are handled uniformly.
+
+#### Batch mode
+
+In addition to our current shutdown logic, we should add a configuration option
+and corresponding conditional for exiting the source once all files have reached
+EOF. This, along with disabled file discovery, would neatly and simply implement
+the oft-requested batch mode.
+
+#### Observability improvements
+
+There's no grand design here, but we should do the work to go through and
+address all of the relevant issues. Some examples:
+
+* Logging when we see that a file is deleted but need to keep it open
+* Not logging noise around small or empty files
+* Optionally silencing errors due to dangling symlinks
+
+## Doc-level Proposal
+
+* Rename `ignore_older` to `skip_older_than`
+* Replace `start_from_beginning` with `ignore_checkpoints` and `read_from` (see
+    earlier section)
+* Replace `oldest_first` with `mode = tail|tail_sequential|batch`
+* Rename `glob_minimum_cooldown` to `discovery_interval` and disable when `mode = batch`
+
+## Rationale
+
+The file source has dramatically overgrown its original design and is causing
+pain for both users and developers of Vector. It warrants spending some time and
+effort to improve both usability and maintainability. Otherwise, we risk losing
+a large and increasing amount of our time to its maintenance and user support.
+
+By focusing on modularity and simple improvements over a rewrite, we will be
+able to maintain the good parts of the source's long history (accumulated bug
+fixes). This also reduces the risk of introducing new bugs that would be
+inevitable with a more aggressive rewrite.
+
+Modularity also positions us well for the future by encouraging a split between
+user-facing config and implementation-level config. This matches well with our
+config composition RFC and should enable simple config facades for specific
+file-based use cases.
+
+## Drawbacks
+
+This is a backwards incompatible change to the config of one of our most widely
+used sources. To change the interface would likely inconvenience a large number
+of current users. There is also always a risk that bugs are introduced as part
+of refactoring, even though our approach is designed to minimize that risk.
```

Comment:
Could we plan a safe deprecation/update path? Eg one version they get a deprecation warning and a pointer to the website how to update, then next version major it's removed?

Reply:
Yes, we can log a warning when we process the old config format for now, and then eventually drop it so people don't have to put `version = 2` for all eternity.